File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] chars = new int[128]; // ASCII size
        int left = 0, right = 0, maxLength = 0;

        while (right < s.length()) {
            char r = s.charAt(right);
            chars[r]++;

            while (chars[r] > 1) {
                char l = s.charAt(left);
                chars[l]--;
                left++;
            }

            maxLength = Math.max(maxLength, right - left + 1);
            right++;
        }

        return maxLength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        count = [0] * 128  
        left = 0
        max_length = 0

        for right in range(len(s)):
            count[ord(s[right])] += 1

            while count[ord(s[right])] > 1:
                count[ord(s[left])] -= 1
                left += 1

            max_length = max(max_length, right - left + 1)

        return max_length
